runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/DoFnOperator.java

```diff
@@ -367,25 +367,50 @@ public void dispose() throws Exception {
       super.dispose();
       checkFinishBundleTimer.cancel(true);
     } finally {
+      if (bundleStarted) {
+        invokeFinishBundle();
+      }
       doFnInvoker.invokeTeardown();
     }
   }
 
   @Override
   public void close() throws Exception {
-    super.close();
+    try {
 
-    // sanity check: these should have been flushed out by +Inf watermarks
-    if (!sideInputs.isEmpty() && nonKeyedStateInternals != null) {
-      BagState<WindowedValue<InputT>> pushedBack =
-          nonKeyedStateInternals.state(StateNamespaces.global(), pushedBackTag);
+      // This is our last change to block shutdown of this operator while
+      // there are still remaining processing-time timers. Flink will ignore pending
+      // processing-time timers when upstream operators have shut down and will also
+      // shut down this operator with pending processing-time timers.
+      while (this.numProcessingTimeTimers() > 0) {
```

Comment:
With a forever-running SDF this is going to block forever, right? Is that ok? It feels icky to have an infinitely long blocking operation in close(), especially given that the operation performs normal work rather than cleanup work.

Reply:
Yes, I also think this is very icky but I don't see another option with how Flink currently works.